airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/task/InputConsumerTask.kt

```diff
@@ -0,0 +1,210 @@
+import io.airbyte.cdk.message.Deserializer
+import io.airbyte.cdk.message.DestinationMessage
+import io.airbyte.cdk.message.DestinationRecord
+import io.airbyte.cdk.message.DestinationRecordWrapped
+import io.airbyte.cdk.message.DestinationStreamAffinedMessage
+import io.airbyte.cdk.message.DestinationStreamComplete
+import io.airbyte.cdk.message.DestinationStreamIncomplete
+import io.airbyte.cdk.message.GlobalCheckpoint
+import io.airbyte.cdk.message.GlobalCheckpointWrapped
+import io.airbyte.cdk.message.MessageQueueSupplier
+import io.airbyte.cdk.message.QueueWriter
+import io.airbyte.cdk.message.StreamCheckpoint
+import io.airbyte.cdk.message.StreamCheckpointWrapped
+import io.airbyte.cdk.message.StreamCompleteWrapped
+import io.airbyte.cdk.message.StreamRecordWrapped
+import io.airbyte.cdk.message.Undefined
+import io.airbyte.cdk.state.MemoryManager
+import io.airbyte.cdk.state.Reserved
+import io.airbyte.cdk.state.SyncManager
+import io.airbyte.cdk.util.use
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Secondary
+import jakarta.inject.Singleton
+import java.io.InputStream
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+
+interface InputConsumerTask : SyncTask
+
+/**
+ * Routes @[DestinationStreamAffinedMessage]s by stream to the appropriate channel and @
+ * [CheckpointMessage]s to the state manager.
+ *
+ * TODO: Handle other message types.
+ */
+@SuppressFBWarnings(
+    "NP_NONNULL_PARAM_VIOLATION",
+    justification = "message is guaranteed to be non-null by Kotlin's type system"
+)
+@Singleton
+@Secondary
+class DefaultInputConsumerTask(
+    private val catalog: DestinationCatalog,
+    private val inputFlow: SizedInputFlow<Reserved<DestinationMessage>>,
+    private val recordQueueSupplier:
+        MessageQueueSupplier<DestinationStream.Descriptor, Reserved<DestinationRecordWrapped>>,
+    private val checkpointQueue: QueueWriter<Reserved<CheckpointMessageWrapped>>,
+    private val syncManager: SyncManager,
+) : InputConsumerTask {
+    private val log = KotlinLogging.logger {}
+
+    private suspend fun handleRecord(
+        reserved: Reserved<DestinationStreamAffinedMessage>,
+        sizeBytes: Long
+    ) {
+        val stream = reserved.value.stream
+        val manager = syncManager.getStreamManager(stream)
+        val queue = recordQueueSupplier.get(stream)
+        when (val message = reserved.value) {
+            is DestinationRecord -> {
+                val wrapped =
+                    StreamRecordWrapped(
+                        index = manager.countRecordIn(),
+                        sizeBytes = sizeBytes,
+                        record = message
+                    )
+                queue.publish(reserved.replace(wrapped))
+            }
+            is DestinationStreamComplete,
+            is DestinationStreamIncomplete -> {
+                reserved.release() // safe because multiple calls conflate
```

Comment:
actually - why do we need to reserve memory at all for these messages? it's not like we actually hold the original message in memory (and even if we did, its a pretty small message anyway)

Reply:
In practical terms, because we reserve in the input flow, when we aren't working with message type yet... though I guess we could check whether it's a record and only reserve for that, but then the type of the flow would be crazy

Reply:
ah yeah, that makes sense. Let's just leave it as-is then :shrug: